src/lib.rs

```diff
@@ -1136,14 +1143,45 @@ impl<T> Drop for Async<T> {
     }
 }
 
-impl<T: Read> AsyncRead for Async<T> {
+/// Types whose I/O trait implementations do not drop the underlying I/O source.
+///
+/// # Safety
+///
+/// Any I/O trait implementations for this type must not drop the underlying I/O source.
+pub unsafe trait IoSafe {}
+
+// Reference types can't be mutated.
```

Comment:
what about mutex types and such?

Reply:
Since it's through an `&` I don't think interior mutability can invalidate any existing `AsFd` references. I've added documentation clarifying this.

Reply:
huh but interior mutability is exactly about being able to invalidate existing inner contents (or at least modifying them concurrently so we can't trust them); I think these "&" impls should require that `T: IoSafe`.

Reply:
The thing is that it's impossible to return a file descriptor-wielding resource from within any current interior mutability container in Rust. This code doesn't work for example:

```rust
struct Foo(RefCell<TcpStream>);

impl AsFd for Foo {
    fn as_fd(&self) -> BorrowedFd<'_> {
        self.0.borrow().as_fd()
        // ^ borrow checker throws an error on this line
    }
}
```

This is because the borrow checker  recognizes that the contents of `foo.0` can be changed during the lifetime of the `BorrowedFd` and thus prevents it from passing. The only primitive where this kind of thing would work is `OnceCell`, and the main reason why it works is because `OnceCell` can't be mutated after the value is first instantiated.

Pinning guarantees operate on generally the same principle.

The worst thing that can happen is that external state is used to change what kind of pointer `as_fd()` returns. For instance:

```rust
struct Bar {
    flag: Cell<bool>,
    a: TcpStream,
    b: TcpStream
}

impl AsFd for Bar {
    fn as_fd(&self) -> BorrowedFd<'_> {
        if self.flag.replace(!self.flag.get()) {
            &self.a
        } else {
            &self.b
        }
    }
}
```

We solve this problem by only calling `as_fd()` once to get the original source. Implementations like this are considered buggy (but not unsound) and are thus not really supported by `async-io`.

Reply:
"The worst thing that can happen" is exactly what I meant. Ok. (imo the last paragraph of your response should be added as a comment into the source)